File: 416. Partition Equal Subset Sum/main.py
from typing import List


class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        if sum(nums) % 2:
            return False
        dp = set()
        dp.add(0)
        target = sum(nums) // 2
        for i in range(len(nums)):
            nextDP = set(dp)
            for t in dp:
                nextDP.add(t + nums[i])
            dp = nextDP
        return target in dp

File: 416. Partition Equal Subset Sum/main.go
package main

func canPartition(nums []int) bool {
	total := 0
	for _, n := range nums {
		total += n
	}
	if total%2 != 0 {
		return false
	}

	target := total / 2
	dp := make([]bool, target+1)
	dp[0] = true

	for _, n := range nums {
		for j := target; j >= n; j-- {
			if dp[j-n] {
				dp[j] = true
			}
		}
	}
	return dp[target]
}
